File: backend/pkg/canonical/canonical.go
// Package canonical provides canonical JSON utilities for FLK/PCS signature generation.
// Phase 10 WP1: Ensures cross-language signature compatibility (Python/Go/TS).
//
// Key requirements:
// - Floats formatted to exactly 9 decimal places
// - Stable field ordering for signature subset
// - No whitespace in JSON output
package canonical

import (
	"encoding/json"
	"fmt"
	"strconv"
)

// SignatureFields defines the 8-field subset used for signing (Phase 1 specification)
var SignatureFields = []string{
	"D_hat",
	"budget",
	"coh_star",
	"epoch",
	"merkle_root",
	"pcs_id",
	"r",
	"shard_id",
}

// SignatureSubset represents the minimal PCS fields used for signing
type SignatureSubset struct {
	PCSID      string  `json:"pcs_id"`
	MerkleRoot string  `json:"merkle_root"`
	Epoch      int     `json:"epoch"`
	ShardID    string  `json:"shard_id"`
	DHat       float64 `json:"D_hat"`
	CohStar    float64 `json:"coh_star"`
	R          float64 `json:"r"`
	Budget     float64 `json:"budget"`
}

// F9 formats a float64 to exactly 9 decimal places.
//
// This ensures stable signatures across languages and prevents
// floating-point drift.
//
// Example:
//
//	F9(1.23456789012345) // returns "1.234567890"
//	F9(0.5)              // returns "0.500000000"
func F9(x float64) string {
	return strconv.FormatFloat(x, 'f', 9, 64)
}

// Round9 rounds a float64 to 9 decimal places.
//
// Used for normalizing floats before formatting to ensure
// consistent behavior across operations.
func Round9(x float64) float64 {
	// Multiply by 10^9, round, divide by 10^9
	const factor = 1e9
	return float64(int64(x*factor+0.5)) / factor
}

// CanonicalJSONBytes generates canonical JSON bytes for signing.
//
// Rules:
//   - Signature fields only (8 fields)
//   - Floats formatted to 9 decimal places
//   - Keys sorted alphabetically
//   - No whitespace (compact JSON)
//   - UTF-8 encoded
//
// Args:
//
//	subset: SignatureSubset with required fields
//
// Returns:
//
//	Canonical JSON as bytes, ready for signing
func CanonicalJSONBytes(subset *SignatureSubset) ([]byte, error) {
	// Normalize floats to 9 decimal places
	normalized := map[string]interface{}{
		"D_hat":       Round9(subset.DHat),
		"budget":      Round9(subset.Budget),
		"coh_star":    Round9(subset.CohStar),
		"epoch":       subset.Epoch,
		"merkle_root": subset.MerkleRoot,
		"pcs_id":      subset.PCSID,
		"r":           Round9(subset.R),
		"shard_id":    subset.ShardID,
	}

	// Marshal with no indentation (compact)
	// Go's json.Marshal already sorts keys alphabetically
	return json.Marshal(normalized)
}

// SignaturePayload extracts signature subset and generates canonical payload.
//
// This is the main entry point for generating bytes to sign.
//
// Example:
//
//	subset := &SignatureSubset{
//	    PCSID:      "test",
//	    DHat:       1.23456789,
//	    CohStar:    0.75,
//	    R:          0.5,
//	    Budget:     0.35,
//	    MerkleRoot: "abc",
//	    Epoch:      1,
//	    ShardID:    "s1",
//	}
//	payload, err := SignaturePayload(subset)
func SignaturePayload(subset *SignatureSubset) ([]byte, error) {
	if subset.PCSID == "" {
		return nil, fmt.Errorf("missing required field: pcs_id")
	}
	if subset.MerkleRoot == "" {
		return nil, fmt.Errorf("missing required field: merkle_root")
	}
	if subset.ShardID == "" {
		return nil, fmt.Errorf("missing required field: shard_id")
	}

	return CanonicalJSONBytes(subset)
}

// ExtractSignatureSubset extracts signature fields from a PCS map.
//
// This helper function is useful when working with dynamic PCS structures.
func ExtractSignatureSubset(pcs map[string]interface{}) (*SignatureSubset, error) {
	subset := &SignatureSubset{}

	// Extract required fields
	if v, ok := pcs["pcs_id"].(string); ok {
		subset.PCSID = v
	} else {
		return nil, fmt.Errorf("missing or invalid pcs_id")
	}

	if v, ok := pcs["merkle_root"].(string); ok {
		subset.MerkleRoot = v
	} else {
		return nil, fmt.Errorf("missing or invalid merkle_root")
	}

	if v, ok := pcs["epoch"].(float64); ok {
		subset.Epoch = int(v)
	} else if v, ok := pcs["epoch"].(int); ok {
		subset.Epoch = v
	} else {
		return nil, fmt.Errorf("missing or invalid epoch")
	}

	if v, ok := pcs["shard_id"].(string); ok {
		subset.ShardID = v
	} else {
		return nil, fmt.Errorf("missing or invalid shard_id")
	}

	if v, ok := pcs["D_hat"].(float64); ok {
		subset.DHat = v
	} else {
		return nil, fmt.Errorf("missing or invalid D_hat")
	}

	if v, ok := pcs["coh_star"].(float64); ok {
		subset.CohStar = v
	} else {
		return nil, fmt.Errorf("missing or invalid coh_star")
	}

	if v, ok := pcs["r"].(float64); ok {
		subset.R = v
	} else {
		return nil, fmt.Errorf("missing or invalid r")
	}

	if v, ok := pcs["budget"].(float64); ok {
		subset.Budget = v
	} else {
		return nil, fmt.Errorf("missing or invalid budget")
	}

	return subset, nil
}

File: sdk/ts/src/canonical.ts
/**
 * Canonical JSON utilities for FLK/PCS signature generation.
 * Phase 10 WP1: Ensures cross-language signature compatibility (Python/Go/TS).
 *
 * Key requirements:
 * - Floats formatted to exactly 9 decimal places
 * - Stable field ordering for signature subset
 * - No whitespace in JSON output
 */

/**
 * Signature subset: 8 fields used for signing (Phase 1 specification)
 */
export const SIGNATURE_FIELDS = [
  "pcs_id",
  "merkle_root",
  "epoch",
  "shard_id",
  "D_hat",
  "coh_star",
  "r",
  "budget",
] as const;

/**
 * Signature subset interface
 */
export interface SignatureSubset {
  pcs_id: string;
  merkle_root: string;
  epoch: number;
  shard_id: string;
  D_hat: number;
  coh_star: number;
  r: number;
  budget: number;
}

/**
 * Format float to exactly 9 decimal places.
 *
 * This ensures stable signatures across languages and prevents
 * floating-point drift.
 *
 * @param value - Float to format
 * @returns String representation with 9 decimal places (e.g., "1.234567890")
 *
 * @example
 * ```typescript
 * formatFloat9dp(1.23456789012345); // returns "1.234567890"
 * formatFloat9dp(0.5);              // returns "0.500000000"
 * ```
 */
export function formatFloat9dp(value: number): string {
  return value.toFixed(9);
}

/**
 * Round float to 9 decimal places.
 *
 * Used for normalizing floats before formatting to ensure
 * consistent behavior across operations.
 *
 * @param value - Float to round
 * @returns Float rounded to 9 decimal places
 */
export function round9(value: number): number {
  const factor = 1e9;
  return Math.round(value * factor) / factor;
}

/**
 * Extract the 8-field subset used for signing.
 *
 * @param pcs - Full PCS object
 * @returns Object with only signature fields
 * @throws Error if required signature field is missing
 */
export function extractSignatureSubset(pcs: Record<string, any>): SignatureSubset {
  const subset: Partial<SignatureSubset> = {};

  for (const field of SIGNATURE_FIELDS) {
    if (!(field in pcs)) {
      throw new Error(`Missing required signature field: ${field}`);
    }
    subset[field as keyof SignatureSubset] = pcs[field];
  }

  return subset as SignatureSubset;
}

/**
 * Generate canonical JSON bytes for signing.
 *
 * Rules:
 * - Signature fields only (8 fields)
 * - Floats formatted to 9 decimal places
 * - Keys sorted alphabetically
 * - No whitespace (compact JSON)
 * - UTF-8 encoded
 *
 * @param pcsSubset - PCS object (should contain signature fields only)
 * @returns Canonical JSON as Uint8Array, ready for signing
 *
 * @example
 * ```typescript
 * const pcs = {
 *   pcs_id: "abc123",
 *   D_hat: 1.23456789,
 *   coh_star: 0.75,
 *   r: 0.5,
 *   budget: 0.35,
 *   merkle_root: "def456",
 *   epoch: 1,
 *   shard_id: "shard-001"
 * };
 * const payload = canonicalJSONBytes(pcs);
 * ```
 */
export function canonicalJSONBytes(pcsSubset: SignatureSubset): Uint8Array {
  // Normalize floats to 9 decimal places
  const normalized: Record<string, any> = {};

  for (const [key, value] of Object.entries(pcsSubset)) {
    if (typeof value === "number" && !Number.isInteger(value)) {
      // Round to 9dp, then parse back to number to preserve precision
      normalized[key] = parseFloat(formatFloat9dp(round9(value)));
    } else {
      normalized[key] = value;
    }
  }

  // Sort keys alphabetically
  const sortedKeys = Object.keys(normalized).sort();
  const sortedObj: Record<string, any> = {};
  for (const key of sortedKeys) {
    sortedObj[key] = normalized[key];
  }

  // Generate compact JSON (no whitespace)
  const jsonStr = JSON.stringify(sortedObj);

  // Convert to UTF-8 bytes
  const encoder = new TextEncoder();
  return encoder.encode(jsonStr);
}

/**
 * Generate signature payload from full PCS.
 *
 * Convenience function that combines subset extraction and
 * canonical JSON generation.
 *
 * @param pcs - Full PCS object
 * @returns Canonical payload bytes for signing
 *
 * @example
 * ```typescript
 * const pcs = {
 *   pcs_id: "test",
 *   D_hat: 1.0,
 *   coh_star: 0.75,
 *   r: 0.5,
 *   budget: 0.35,
 *   merkle_root: "abc",
 *   epoch: 1,
 *   shard_id: "s1",
 *   extra_field: "ignored"
 * };
 * const payload = signaturePayload(pcs);
 * ```
 */
export function signaturePayload(pcs: Record<string, any>): Uint8Array {
  const subset = extractSignatureSubset(pcs);
  return canonicalJSONBytes(subset);
}

/**
 * Backwards compatibility: Phase 1-9 used this function name.
 * @deprecated Use signaturePayload() instead.
 */
export function SigningPayload(pcs: Record<string, any>): Uint8Array {
  return signaturePayload(pcs);
}
